### ISSUE TITLE
K8S client issues

### ISSUE THREAD

**Issue by alacuku** [falcosecurity/falco#2973]
**Motivation**

Falco has a built-in functionality called **Kubernetes Metadata Enrichment**. It provides k8s metadata, fetched from the k8s **api-server**, used by Falco to enrich the **system-call events**. Furthermore, these metadata are available to users as **[events fields](https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s)** to be used in the `conditions` and 'outputs' of Falco rules.

## Current implementation

The built-in k8s client is written in C++ and lives in the [libs repo](https://github.com/falcosecurity/libs). It has a number of issues discussed in the next sections.

### Poor performances
The **metadata** are fetched asynchronously from the **api-server** by opening a connection to the `watch` endpoint. Periodically, it checks for new events and processes them. The processing is done in the main loop, bloking the processing of *syscall* stream coming from the driver/probe. In large environments this causes syscall events to be dropped. For more info, please see: https://github.com/falcosecurity/falco/issues/2129. The following scheme shows how the client handles events coming from the **api-server** (some calls have been ommitted for the sake of clarity):

![built-in-k8s-client-flow](https://user-images.githubusercontent.com/17413176/223686298-6b52e081-b932-468e-a474-e3e45ec692c8.svg)

### Stability issues
The network connections to the **api-server** are far from being stable. It does not check for temporary network failures causing Falco to crash and restart. When deploying Falco in large clusters, all the Falco instances( one per node) connects to the **api-server**, and it could happen that some of them are **throtled** and the connection is closed. In such cases Falco pods will crash and restart. For more info see: https://github.com/falcosecurity/falco/issues/1909

### Scalability issues
A **watch** is opened toward the **api-server**for each resource. Currently, we collect metadata for the following resources:
* pods;
* deployments;
* replictioncontrollers;
* replicasets;
* daemonsets;
* namespaces;
* services;
* nodes.

In a 5K nodes cluster the Falco instances will open 40000 watches. This approach does not scale well and contributes to control plane outages in large clusters. 

Only Pods objects are efficiently retrieved by using the NodeSelector. For the other resource types, there is no way to filter them out, but we cache them locally leading to a significant waste of resources, especially memory.

Known issues:
* https://github.com/falcosecurity/falco/issues/1909
* https://github.com/falcosecurity/falco/issues/2346
* https://github.com/falcosecurity/falco/issues/2129
* https://github.com/falcosecurity/falco/issues/1781
* https://github.com/falcosecurity/falco/issues/2463
* https://github.com/falcosecurity/falco/issues/2975

## Native libs (client-go) based implementation

The idea is to introduce a centralized component that retrieves the K8s metadata required by Falco instances. The new component should be based on modern libraries such as the one provided by the k8s project. 

### Motivation

See the section about current implementation.

### Goals

* Fetch metadata from the k8s `api-server` using a centralized component;
* Keep the latency of the updates as short as possible;
* Send only the required metadata to Falco instances ready to be used without the need for further processing on Falco's side;

### Non-Goals

* Sharding the K8s state, the instance of the component will be responsible for the entire cluster;
* Adaptive metadata fetching based on requests by the Falco instances. We could think about it in the next iterations.

### Proposal

Provide a new component to be deployed in the K8s cluster responsible for providing the `metadata` required by each Falco instance. `Metadata` to be sent to Falco:

* Pod's metadata for a given node;
* Namespace metadata only for the ones where local pods reside;
*  Workloads metadata: Deployment, replicasets, daemonsets, replicationcontrollers;
* Services metadata only for the ones serving the local pods;

Each Falco instance will subscribe to the `centralized component` providing the `node name` where they reside and will receive metadata through a stream of updates. The communication between the Falco instances and the `centralized component` uses GRPC over HTTP2.

![Image](https://user-images.githubusercontent.com/17413176/226294705-2553c173-7c3d-4f85-bbe5-931e567e97c9.png)

Design Details

The **Meta Collector** will be implemented using Go relying on [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) to efficiently retrieve and cache K8s resources. An `operator/controller` will be implemented for each resource needed by Falco. Each operator will connect to the `api-server` and fetch the initial state for the watched resource. It will create a state in memory and send updates to the Falco instances that need the data. Since we are limited to a subset of the fields of the watched resources update events will be sent only when those fields change. Updates should be sent to Falco as soon as possible. Each operator will:
* Check if a `Modified` event received from the `api-server` produces an `update` to be sent to the Falco instances;
* Generate the event to be sent;
* Send the event to the `message broker` which will send it to the relevant Falco instances.

![Image](https://user-images.githubusercontent.com/17413176/226299056-7b8dcdc2-b5b1-429a-a0f0-86c8217b1816.png)

After the `operators` complete the initialization a GRPC server (message broker) will start and wait for Falco subscriptions. The message broker will compute an initial list of `events/updates` to send to the Falco instance upon a new subscription. Remember that the state is cached, so subscriptions will not generate requests to the `api-server`. Then the `updates` will be sent to the Falco instances as they come from the `api-server`.

**NOTE** This issue will be updated with more implementation details in the upcoming weeks.

**Comment by Andreagit97** [comment]
> so I hope to simply keep the syscall pods separate from the API server pods as it will greatly simplify the implementation of these (due to security policies in our clusters) for our org once the new components are released.

yes, this is the idea, we will see if during the integration we will face some issues in doing this. We will keep you posted!

**Comment by Issif** [comment]
I like the principle too. 
I already discussed about that with @leogr  @LucaGuerra and @alacuku, but this central component could be extended in the future to enrich the events with fields from the cloud provider metadata or just static elements (like falcosidekick does).
Moreover, it will imply that Falco inits the gRPC connection to the  metadata-collector, right now, it's the opposite for the falco-exporter. I guess it could imply in the future the same direction for the output, and falco could be connected to sidekick or else with gRPC.

**Comment by erfanw** [comment]
Do we have an estimated time when this new implementation will be available?
Thank you for your work!

**Comment by leogr** [comment]
> Do we have an estimated time when this new implementation will be available? Thank you for your work!

Alpha versions of the [plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta) and the [meta-collector](https://github.com/falcosecurity/k8s-metacollector) are already available, but they are still under active development.

We aim to officially release them (still marked as experimental) close to Falco release 0.37 (due by the end of Jan).

**Comment by alacuku** [comment]
The alpha versions of the [k8s-metacollector](https://github.com/falcosecurity/k8s-metacollector) and the [k8s-meta plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta) are out. 

There is also a pull request https://github.com/falcosecurity/charts/pull/598 that integrates the new components with Falco `0.36.2` and makes it easy to test. In the pull request description, you will find anything you need to test it.

@Dentrax @tspearconquest @erfanw

**Comment by alacuku** [comment]
> Hi, as this is to be a few separate components, and it will be talking with the API server, I wanted to ask to consider putting these components into a separate namespace from the main falco pods. Better still would be a separate namespace for each component, but I realize that the single shared namespace is a common pattern, so I hope to simply keep the syscall pods separate from the API server pods as it will greatly simplify the implementation of these (due to security policies in our clusters) for our org once the new components are released.

We developed a new [chart](https://github.com/falcosecurity/charts/tree/master/charts/k8s-metacollector)  for the `k8s-metacollector`. The `Falco` chart declares it as a dependency and allows to override the namespace for the dependent charts:
```bash
# First create the custom namespace for the collector.
kubectl create ns collector
# Install falco + collector.
helm install falco  falcosecurity/falco
    -n falco \
    --create-namespace \
    --set k8s-metacollector.namespaceOverride=collector
```

**Comment by incertum** [comment]
@alacuku prior to the new k8smeta plugin the helm chart had `-k` flag set by default, but now I hope we do not enable the new plugin by default right?

**Comment by alacuku** [comment]
The chart has the same behavior as before: by default the `k8s-metacollector` and `k8smeta` plugin are deployed by the helm chart. Users that do not need it, can disable it by setting `collectors.kubernetes.enabled=false`.

**Comment by incertum** [comment]
@falcosecurity/falco-maintainers would you all be open to reconsider enabling `k8s-metacollector` and `k8smeta` by default? We should have a formal vote on this matter.

The reasons being that historically `-k` has caused a tremendous amount of issues and it wasn't even used by Falco by default. We lack the data to ensure the new `k8smeta` plugin is more performant. 
However, more importantly the default Falco deployment does not use any of the new `k8smeta` fields, e.g. see the stable rules, they only use the k8s fields retrieved from the container runtime socket when running Falco with `-pk` (resolving the placeholder `%container.info` with the namespace and pod name). 

In general, I believe no plugins should be enabled by default in any templates or default configs we provide. Each plugin should be strictly opt-in. Because enabling one plugin versus disabling another plugin also begs the question as to why? What are the formal criteria to say a plugin should be enabled by default? ... Especially if not needed by the default Falco config it equates to wasted CPU cycles. 
That way we ensure the default configs are the leanest possible and onboarding for newcomers has higher chances of going smooth and the debugging burden on maintainers when helping adopters will also be reduced. 

WDYT?

-1 (my vote) -> do not enable `k8s-metacollector` and `k8smeta` in the helm chart by default

**Comment by alacuku** [comment]
@incertum, calling for a vote on this issue does not bring any benefit to users that are afflicted by the kubernetes client issues. 
Could we please open a new issue, and use this one for updates regarding the new implementation?

**Comment by incertum** [comment]
🙃 fair, here we go created a new issue and moved my post over.

**Comment by mateuszdrab** [comment]
No sure if relevant, but I've been having a different issue with K8s metadata becoming unavailable in the Falco pods after sometime of pod uptime resulting in lots of alerts being raised as process metadata are not available (e.q pod name and namespace) which results in the alert rule exclusions no longer apply to them. 
The issue is resolved by a restart of the pod.

Could it be related to this? If not, I'm happy to raise a separate issue for this.

**Comment by Andreagit97** [comment]
> Could it be related to this? If not, I'm happy to raise a separate issue for this.
ei @mateuszdrab yes it  could be related to this

Have you tried this one by chance https://github.com/falcosecurity/falco/issues/2973#issuecomment-1876889214?
This is a new way of collecting k8s-metadata that will be introduced in Falco 0.37.0 (release date ~ next week). It should solve the issue!
the link above is a sort of preview helm chart that works with Falco 0.36.2, if you want to try it!

**Comment by mateuszdrab** [comment]
Hi @Andreagit97
I'll try it if time permits, if not, it will get deployed automatically once 0.37 is out

**Comment by alacuku** [comment]
> Hi @Andreagit97 I'll try it if time permits, if not, it will get deployed automatically once 0.37 is out

@mateuszdrab, keep in mind that the new `k8s-metacollector` will be disabled by default.

**Comment by pirxthepilot** [comment]
Hi @alacuku , just to clarify, in 0.37.0 we can still use the old k8s-client, it doesn't entirely deprecate it?

**Comment by alacuku** [comment]
Hi @pirxthepilot, the old k8s-client will be removed in falco `0.37.0`. But the [k8s.*](https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s) are still available. Most of them are stubbed, and will return `N/A` but the `k8s.pod.*` and `k8s.ns.name` fields are populated with data gathered from the container runtime. 

If you need other fields, see here: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields

**Comment by mateuszdrab** [comment]
I think I'm a bit confused now. 
The k8s client is being removed and replaced by k8smeta and k8s-metacollector? 
By default, both will be disabled? (As per https://github.com/falcosecurity/falco/issues/2995)
I use fields like `container.image.repository`, `container.id` and `k8s.ns.name` and don't think I need the new ones; however, I do want to address the issue of metadata being suddenly lost resulting in an increase in alerts like I mentioned above.

**Comment by leogr** [comment]
Hey @mateuszdrab 

> I think I'm a bit confused now. 

You're right, this may be confusing. I guess the lack of clarity is due to the missing documentation (we are working on it right now). Meanwhile, I can give you an overview, hoping it will help with make more clarity.

> The k8s client is being removed and replaced by k8smeta and k8s-metacollector? 
Yes.

> By default, both will be disabled? (As per #2995) 
Yes, they will. Please note that you can see k8smeta and k8s-metacollector as two components of a single feature. The k8smeta does not work without the k8s-metacollector. So let's say these two components together are the new **_advanced** k8s metadata enrichment_ mechanism for Falco (below it will clear why I put emphasis on the **_advanced** word.

> I use fields like `container.image.repository`, `container.id` and `k8s.ns.name` and don't think I need the new ones; 

All `container.*`,  `k8s.pod.*` and `k8s.ns.name` fields are populated with data gathered from the container runtime. So they work out-of-the-box in Falco without the need for the _old k8s client_ nor the _new advanced k8s metadata enrichment_. Thus, these fields will continue to work as before. Nothing changed on this front.


> however, I do want to address the issue of metadata being suddenly lost resulting in an increase in alerts like I mentioned above.

Starting with Falco 0.37 (due by Jan 29th), you will be able to enable the new advanced k8s metadata enrichment (k8smeta and k8s-metacollector) as well. 
The new fields provided by the new advanced k8s metadata enrichment are in addition and do not replace the existing `k8s.pod.*` and `k8s.ns.name` fields. This also implies you will be able to use both `k8s.pod.name` and `k8smeta.pod.name` at the same time. In most cases, they will return the same value, but since the data is collected from different places (the container runtime for `k8s` ones, the K8s API server, for `k8smeta`), they availability and reliability can be different during the lifecycle of an application. Although this solution might seem redundant, it will offer a lot of flexibility to the users. In addition, the new k8s metadata enrichment will provide a lot of other fields (you can find them here: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields). Note that the new advanced k8s metadata enrichment requires a cluster component (i.e. the k8smetacollector) to be deployed in the cluster, so it might not fit all users.

For these reasons, and since the new advanced k8s metadata enrichment might not be required by all users, we consider it an "advanced" mechanism, and it is not enabled by default.

To wrap up:
- If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are enough for your use case, you should evaluate and stick with them.
- If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are not enough for you (because they do not work well in your use case or you need more metadata) you should evaluate to opt-in to the new k8s metadata enrichment mechanism. This will require a bit of experimentation (since it's a fresh new approach), but you can do it in parallel.


Hoping this helps. Let me know if you have any other doubts!

**Comment by mateuszdrab** [comment]
> Hey @mateuszdrab 
> 
> > I think I'm a bit confused now. 
> 
> You're right, this may be confusing. I guess the lack of clarity is due to the missing documentation (we are working on it right now). Meanwhile, I can give you an overview, hoping it will help with make more clarity.
> 
> > The k8s client is being removed and replaced by k8smeta and k8s-metacollector? 
> Yes.
> 
> > By default, both will be disabled? (As per #2995) 
> Yes, they will. Please note that you can see k8smeta and k8s-metacollector as two components of a single feature. The k8smeta does not work without the k8s-metacollector. So let's say these two components together are the new **_advanced** k8s metadata enrichment_ mechanism for Falco (below it will clear why I put emphasis on the **_advanced** word.
> 
> > I use fields like `container.image.repository`, `container.id` and `k8s.ns.name` and don't think I need the new ones; 
> 
> All `container.*`,  `k8s.pod.*` and `k8s.ns.name` fields are populated with data gathered from the container runtime. So they work out-of-the-box in Falco without the need for the _old k8s client_ nor the _new advanced k8s metadata enrichment_. Thus, these fields will continue to work as before. Nothing changed on this front.
> 
> 
> > however, I do want to address the issue of metadata being suddenly lost resulting in an increase in alerts like I mentioned above.
> 
> Starting with Falco 0.37 (due by Jan 29th), you will be able to enable the new advanced k8s metadata enrichment (k8smeta and k8s-metacollector) as well. 
> The new fields provided by the new advanced k8s metadata enrichment are in addition and do not replace the existing `k8s.pod.*` and `k8s.ns.name` fields. This also implies you will be able to use both `k8s.pod.name` and `k8smeta.pod.name` at the same time. In most cases, they will return the same value, but since the data is collected from different places (the container runtime for `k8s` ones, the K8s API server, for `k8smeta`), they availability and reliability can be different during the lifecycle of an application. Although this solution might seem redundant, it will offer a lot of flexibility to the users. In addition, the new k8s metadata enrichment will provide a lot of other fields (you can find them here: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields). Note that the new advanced k8s metadata enrichment requires a cluster component (i.e. the k8smetacollector) to be deployed in the cluster, so it might not fit all users.
> 
> For these reasons, and since the new advanced k8s metadata enrichment might not be required by all users, we consider it an "advanced" mechanism, and it is not enabled by default.
> 
> To wrap up:
> - If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are enough for your use case, you should evaluate and stick with them.
> - If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are not enough for you (because they do not work well in your use case or you need more metadata) you should evaluate to opt-in to the new k8s metadata enrichment mechanism. This will require a bit of experimentation (since it's a fresh new approach), but you can do it in parallel.
> 
> 
> Hoping this helps. Let me know if you have any other doubts! 
> 
> 
> 
> 
> 
> 
> 
> 
> 
>  

Thank you for clarifying

It looks like the issue I might be having is not going to be addressed by this change. For me, it is those basic K8s.* labels I mentioned that disappear resulting in excessive alerts. I can resolve it by restarting the Falco pods.

I am surprised the K8s namespace metadata comes from the container runtime and not the API server itself.

I guess I may need to raise a separate issue for that if I determine the exact steps to reproduce it.

**Comment by mateuszdrab** [comment]
> Hey @mateuszdrab 
> 
> > I think I'm a bit confused now. 
> 
> You're right, this may be confusing. I guess the lack of clarity is due to the missing documentation (we are working on it right now). Meanwhile, I can give you an overview, hoping it will help with make more clarity.
> 
> > The k8s client is being removed and replaced by k8smeta and k8s-metacollector? 
> Yes.
> 
> > By default, both will be disabled? (As per #2995) 
> Yes, they will. Please note that you can see k8smeta and k8s-metacollector as two components of a single feature. The k8smeta does not work without the k8s-metacollector. So let's say these two components together are the new **_advanced** k8s metadata enrichment_ mechanism for Falco (below it will clear why I put emphasis on the **_advanced** word.
> 
> > I use fields like `container.image.repository`, `container.id` and `k8s.ns.name` and don't think I need the new ones; 
> 
> All `container.*`,  `k8s.pod.*` and `k8s.ns.name` fields are populated with data gathered from the container runtime. So they work out-of-the-box in Falco without the need for the _old k8s client_ nor the _new advanced k8s metadata enrichment_. Thus, these fields will continue to work as before. Nothing changed on this front.
> 
> 
> > however, I do want to address the issue of metadata being suddenly lost resulting in an increase in alerts like I mentioned above.
> 
> Starting with Falco 0.37 (due by Jan 29th), you will be able to enable the new advanced k8s metadata enrichment (k8smeta and k8s-metacollector) as well. 
> The new fields provided by the new advanced k8s metadata enrichment are in addition and do not replace the existing `k8s.pod.*` and `k8s.ns.name` fields. This also implies you will be able to use both `k8s.pod.name` and `k8smeta.pod.name` at the same time. In most cases, they will return the same value, but since the data is collected from different places (the container runtime for `k8s` ones, the K8s API server, for `k8smeta`), they availability and reliability can be different during the lifecycle of an application. Although this solution might seem redundant, it will offer a lot of flexibility to the users. In addition, the new k8s metadata enrichment will provide a lot of other fields (you can find them here: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields). Note that the new advanced k8s metadata enrichment requires a cluster component (i.e. the k8smetacollector) to be deployed in the cluster, so it might not fit all users.
> 
> For these reasons, and since the new advanced k8s metadata enrichment might not be required by all users, we consider it an "advanced" mechanism, and it is not enabled by default.
> 
> To wrap up:
> - If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are enough for your use case, you should evaluate and stick with them.
> - If `container.*`,  `k8s.pod.*`, and `k8s.ns.name` fields are not enough for you (because they do not work well in your use case or you need more metadata) you should evaluate to opt-in to the new k8s metadata enrichment mechanism. This will require a bit of experimentation (since it's a fresh new approach), but you can do it in parallel.
> 
> 
> Hoping this helps. Let me know if you have any other doubts! 
> 
> 
> 
> 
> 
> 
> 
> 
> 
>  

Thank you for clarifying

It looks like the issue I might be having is not going to be addressed by this change. For me, it is those basic K8s.* labels I mentioned that disappear resulting in excessive alerts. I can resolve it by restarting the Falco pods.

I am surprised the K8s namespace metadata comes from the container runtime and not the API server itself.

I guess I may need to raise a separate issue for that if I determine the exact steps to reproduce it.

**Comment by leogr** [comment]
> It looks like the issue I might be having is not going to be addressed by this change. For me, it is those basic K8s.* labels I mentioned that disappear resulting in excessive alerts. I can resolve it by restarting the Falco pods.

This might be related to drops or other issues related to the container runtime. Or perhaps, some conflict with the old k8s client (have you tried to remove -k/-K?). In any case, I agree yours is likely a different issue.

> I guess I may need to raise a separate issue for that if I determine the exact steps to reproduce it.
This would be great :+1: 

> I am surprised the K8s namespace metadata comes from the container runtime and not the API server itself.
Magic :exploding_head: This was introduced a couple of years ago in Falco and is likely not well-documented. I hope we will do better in documenting the feature coming with the new release (cc @LucaGuerra ).

**Comment by Andreagit97** [comment]
This should be solved by [Falco 0.37.0](https://github.com/falcosecurity/falco/releases/tag/0.37.0)! Feel free to reopen if this is still an issue

**Comment by mateuszdrab** [comment]
Hi @Andreagit97 

Since the update, it looks like Falco isn't populating the `k8s.ns.name` and `k8s.pod` fields - they're all set to `null`.
All the rule exceptions that used those fields are not working and result in thousands of alerts being generated.

Restarted the daemonset but the situation continues.
Anything else you'd recommend me check or shall I raise a new issue?

**Comment by Andreagit97** [comment]
ei @mateuszdrab some questions:
* can you provide us with the helm chart command you are using to deploy Falco?
* which Falco version are you using?
* before the update were these fields available?

**Comment by mateuszdrab** [comment]
Hi @Andreagit97 

Values for the deployment are:
```
controller:
  daemonset:
    updateStrategy:
      rollingUpdate:
        maxUnavailable: 100%
customRules:
  rules-custom.yaml: <redacted>
driver:
  enabled: true
  kind: ebpf
falco:
  grpc:
    enabled: true
  grpc_output:
    enabled: true
  rules_file:
  - /etc/falco/falco_rules.yaml
  - /etc/falco/falco-incubating_rules.yaml
  - /etc/falco/falco-sandbox_rules.yaml
  - /etc/falco/rules.d
  syscall_event_drops:
    actions:
    - log
falcoctl:
  config:
    artifact:
      follow:
        refs:
        - falco-rules:2
        - falco-incubating-rules:2
        - falco-sandbox-rules:2
      install:
        refs:
        - falco-rules:2
        - falco-incubating-rules:2
        - falco-sandbox-rules:2
falcosidekick:
  config:
    syslog:
      host: <redacted>
      port: "514"
    teams:
      webhookurl: <redacted>
  enabled: true
  webui:
    enabled: false
tolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
- effect: NoSchedule
  key: node-role.kubernetes.io/control-plane
- key: environment
  operator: Exists
- key: worker-role
  operator: Exists
```

Version is latest so `0.37.1`, and indeed, before update the fields were available so the rules were working fine which can be seen in the screenshot below:

![image](https://github.com/falcosecurity/falco/assets/26556541/f2cc027c-a726-43c5-827d-a2da558f23fe)

In the query below, you can see a sample alert that used to have a namespace in the json fields but now no longer has it
![image](https://github.com/falcosecurity/falco/assets/26556541/a4480c57-9aa4-4aaa-b1e4-b276f857560f)

**Comment by Andreagit97** [comment]
From the config you posted it seems like the Kubernetes collection is disabled -> https://github.com/falcosecurity/charts/blob/68db0665712d657be00be36e5b7933626220d3e7/charts/falco/values.yaml#L289, you should enable it... When Falco starts it should show you a log telling you that the `k8smeta` plugin is injected and ready to receive data. Can you also post the Falco startup logs?

BTW not sure which container runtime you are using but `k8s.pod.*` and `k8s.ns.name` should be available also through the container runtime https://github.com/falcosecurity/charts/blob/falco-4.2.5/charts/falco/values.yaml#L268 so if you don't see these info and you have the container collector enabled there is something strange...can you provide more info about your setup?

If you are searching for more info on how k8s fields are provided i will leave the link here https://falco.org/blog/falco-0-37-0/#fields-supported-by-the-new-k8smeta-plugin

**Comment by mateuszdrab** [comment]
Hey Andrea
I'm running k3s version v1.27.7+k3s2 with containerd as the runtime.

From our previous conversation I thought `k8s.pod.*` and `k8s.ns.name` should be available so I did not enable the new `k8smeta` plugin.

I can see the socket is mounted in `/host/run/containerd/containerd.sock` and it's available; however, I can't see this path being referenced anywhere in Falcos config. Is it hardcoded?

**Comment by Andreagit97** [comment]
> From our previous conversation I thought k8s.pod.* and k8s.ns.name should be available so I did not enable the new k8smeta plugin.

Ok, I updated my previous answer because it was not clear. Yes, `k8s.pod.*` and `k8s.ns.name` fields are available through the container runtime so you should see these... Some additional questions:
* Are these fields available sometimes or are they never filled?
* Which was the previous version of Falco with these fields available? 0.36.0?
* For what concern other `container.*` fields, such as `container.name`, `container.id`, do you see them, or again are they never filled? I would like to understand if the container enrichment doesn't work at all with this setup or if it doesn't work just for k8s fields.

> I can see the socket is mounted in /host/run/containerd/containerd.sock and it's available; however, I can't see this path being referenced anywhere in Falcos config. Is it hardcoded?

the helm chart do the magic under the hood, passing this value through the Falco command line https://github.com/falcosecurity/charts/blob/68db0665712d657be00be36e5b7933626220d3e7/charts/falco/templates/pod-template.tpl#L67

**Comment by mateuszdrab** [comment]
> Are these fields available sometimes or are they never filled?

From what I can see in Grafana, since Feb 6th the field was never populated.

> Which was the previous version of Falco with these fields available? 0.36.0?

Indeed that's correct

> For what concern other container.* fields, such as container.name, container.id, do you see them, or again are they never filled? I would like to understand if the container enrichment doesn't work at all with this setup or if it doesn't work just for k8s fields.

`container.id` is there

> the helm chart do the magic under the hood, passing this value through the Falco command line https://github.com/falcosecurity/charts/blob/68db0665712d657be00be36e5b7933626220d3e7/charts/falco/templates/pod-template.tpl#L67

That's weird because the socket is mounted in `/host/run/containerd/containerd.sock` but the `--cri` argument in Falco is pointing to `/run/containerd/containerd.sock`. I tried changing it and also added `--disable-cri-async` but no luck.

**Comment by Andreagit97** [comment]
> `container.id` is there

Okay this is fine because it's not extracted directly from the container runtime, what about the `container.name` or `container.image`?
 
> That's weird because the socket is mounted in `/host/run/containerd/containerd.sock` but the `--cri` argument in Falco is pointing to `/run/containerd/containerd.sock`. 

Yep this is fine because internally Falco will add `/host` to the provided path, so the result will be `host/run/containerd/containerd.sock`

**Comment by mateuszdrab** [comment]
> 
> > `container.id` is there
> 
> Okay this is fine because it's not extracted directly from the container runtime, what about the `container.name` or `container.image`?
>  
> > That's weird because the socket is mounted in `/host/run/containerd/containerd.sock` but the `--cri` argument in Falco is pointing to `/run/containerd/containerd.sock`. 
> 
> Yep this is fine because internally Falco will add `/host` to the provided path, so the result will be `host/run/containerd/containerd.sock`
> 
> 
> 

That makes sense, and no, the container image and name are missing.

Example log line (sorry I posted it above but it was hard to notice under the Falco startup warnings):

```
{"hostname":"falco-4cvkq","output":"15:15:10.823823828: Notice Unexpected UDP Traffic Seen (connection=10.42.8.48:35662->216.58.215.68:65535 lport=65535 rport=35662 fd_type=ipv4 fd_proto=fd.l4proto evt_type=connect user=haproxy user_uid=1000 user_loginuid=-1 process=haproxy proc_exepath=/usr/local/sbin/haproxy parent=haproxy-ingress command=haproxy -f /etc/haproxy/haproxy.cfg.3dbbea03-1ebe-4d0a-876c-441ac4f102ab -c -f /etc/haproxy/haproxy-aux.cfg terminal=0 exe_flags=<NA> container_id=69d22f20d812 container_image=<NA> container_image_tag=<NA> container_name=<NA> k8s_ns=<NA> k8s_pod_name=<NA>)","priority":"Notice","rule":"Unexpected UDP Traffic","source":"syscall","tags":["TA0011","container","host","maturity_incubating","mitre_exfiltration","network"],"time":"2024-03-18T15:15:10.823823828Z", "output_fields": {"container.id":"69d22f20d812","container.image.repository":null,"container.image.tag":null,"container.name":null,"evt.arg.flags":null,"evt.time":1710774910823823828,"evt.type":"connect","fd.lport":65535,"fd.name":"10.42.8.48:35662->216.58.215.68:65535","fd.rport":35662,"fd.type":"ipv4","k8s.ns.name":null,"k8s.pod.name":null,"proc.cmdline":"haproxy -f /etc/haproxy/haproxy.cfg.3dbbea03-1ebe-4d0a-876c-441ac4f102ab -c -f /etc/haproxy/haproxy-aux.cfg","proc.exepath":"/usr/local/sbin/haproxy","proc.name":"haproxy","proc.pname":"haproxy-ingress","proc.tty":0,"user.loginuid":-1,"user.name":"haproxy","user.uid":1000}}
```

**Comment by Andreagit97** [comment]
Ok, thank you for all the info! yes, this seems a regression on `k3s v1.27.7+k3s2 with containerd` since Falco 0.36.0 worked on this same env, we will take a look! (@incertum any idea? you probably have more experience with container runtime)

As a temp workaround, you could try to enable the `k8smeta` plugin to collect the k8s metadata, as explained here https://github.com/falcosecurity/falco/issues/2973#issuecomment-2004102708

**Comment by mateuszdrab** [comment]
Sure @Andreagit97 

Just did a quick (slow) Loki query over the past 90 days and we can see the below

![image](https://github.com/falcosecurity/falco/assets/26556541/838887f2-148a-412e-bcce-ace5b0a377a1)

Since Feb 6th, there have been thousands of alerts with no namespace. 
Before that, there were very little alerts because they were being suppressed successfully.
You can also see a spike of alerts before Feb 6th, this was actually the reason I started this investigation which I mentioned earlier https://github.com/falcosecurity/falco/issues/2973#issuecomment-1902765203. 
Essentially, there used to be periods when Falco would no longer respect my overrides, turned out the pods had to be restarted as the k8s fields were missing.

**Comment by Andreagit97** [comment]
Ok got it and Feb 6th is the date on which you switched from 0.36.0 to 0.37.1... pretty interesting

> You can also see a spike of alerts before Feb 6th, this was actually the reason I started this investigation which I mentioned earlier https://github.com/falcosecurity/falco/issues/2973#issuecomment-1902765203.
Essentially, there used to be periods when Falco would no longer respect my overrides, turned out the pods had to be restarted as the k8s fields were missing.

Unfortunately this seems yet another issue :/

**Comment by mateuszdrab** [comment]
> `k8smeta`

True, let's focus on this one for now as it's not working at all.

I've enabled `k8smeta` and I'm still not seeing any metadata added.

I'm upgrading k3s now to v1.27.11-k3s1 - will report back soon.

**Comment by Andreagit97** [comment]
Please note that when you enable the `k8smeta` plugin the name of the fields to use is different, so no more `k8s.pod.*` but you need to use `k8smeta.pod.*` so probably you need to change these fields in your rules

**Comment by alacuku** [comment]
@mateuszdrab here you can find all the fields exposed by the k8smeta plugin: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields

**Comment by Andreagit97** [comment]
I can confirm Falco 0.37.1 doesn't extract `k8s*` fields on K3s, while Falco 0.36.0 seems to do the job. Both versions don't extract the `container.name` and `container.image` maybe we could evaluate exposing all this info through the `k8smeta` plugin and use it as the standard collection method in k8s environments.

**Comment by mateuszdrab** [comment]
> @mateuszdrab here you can find all the fields exposed by the k8smeta plugin: https://github.com/falcosecurity/plugins/tree/master/plugins/k8smeta#supported-fields

Unfortunately, despite enabling the plugin, no `k8smeta` labels are present.

**Comment by incertum** [comment]
Question: Talking about k3s shouldn't the socket be located at `/run/k3s/containerd/containerd.sock` instead? If not too much trouble we would appreciate feedback on the above.

For example the CNCF Green Reviews TAG has k3s containerd on their cluster and I am getting the container information. See the setup here: We explicitly mount the k3s path as well https://github.com/falcosecurity/cncf-green-review-testing/blob/main/kustomize/falco-driver/modern_ebpf/daemonset.yaml#L77 and https://github.com/falcosecurity/cncf-green-review-testing/blob/main/kustomize/falco-driver/modern_ebpf/daemonset.yaml#L124 plus it's one of the default locations, see our Troubleshooting Guide https://falco.org/docs/troubleshooting/missing-fields/#missing-container-images therefore no need to pass the socket location with the `--cri` flag unless it's a custom socket location after all.

**Comment by mateuszdrab** [comment]
That was it, the `/run/containerd/containerd.sock` belongs to Docker's containerd on the same node. 
K3s runs it's own containerd in `/run/k3s/containerd/containerd.sock`.

I was checking the paths earlier and I was trying to remember how it was set up since I've not touched it in ages.

It remains bizzare how it used to work and now it doesn't.

Thank you @incertum!

Maybe worth adding some sort of validation at start-up that the socket is working but has no pods or something like that?

Btw, k8smeta collector still isn't populating its fields
```
{
    "hostname": "falco-6hjsd",
    "output": "22:18:09.254710323: Notice Unexpected UDP Traffic Seen (connection=10.42.8.141:35588->104.16.60.181:443 lport=35588 rport=443 fd_type=ipv4 fd_proto=fd.l4proto evt_type=connect user=root user_uid=0 user_loginuid=-1 process=stats_thread proc_exepath=/fr24feed/fr24feed/fr24feed parent=bash command=stats_thread terminal=0 exe_flags=<NA> container_id=a5613d461804 container_image=docker.io/thomx/fr24feed-piaware container_image_tag=latest container_name=fr24feed-piaware k8s_ns=flight-tracking k8s_pod_name=fr24feed-piaware-ashford-67f544d679-b2qbx)",
    "priority": "Notice",
    "rule": "Unexpected UDP Traffic",
    "source": "syscall",
    "tags": [
        "TA0011",
        "container",
        "host",
        "maturity_incubating",
        "mitre_exfiltration",
        "network"
    ],
    "time": "2024-03-19T22:18:09.254710323Z",
    "output_fields": {
        "container.id": "a5613d461804",
        "container.image.repository": "docker.io/thomx/fr24feed-piaware",
        "container.image.tag": "latest",
        "container.name": "fr24feed-piaware",
        "evt.arg.flags": null,
        "evt.time": 1710886689254710323,
        "evt.type": "connect",
        "fd.lport": 35588,
        "fd.name": "10.42.8.141:35588->104.16.60.181:443",
        "fd.rport": 443,
        "fd.type": "ipv4",
        "k8s.ns.name": "flight-tracking",
        "k8s.pod.name": "fr24feed-piaware-ashford-67f544d679-b2qbx",
        "proc.cmdline": "stats_thread",
        "proc.exepath": "/fr24feed/fr24feed/fr24feed",
        "proc.name": "stats_thread",
        "proc.pname": "bash",
        "proc.tty": 0,
        "user.loginuid": -1,
        "user.name": "root",
        "user.uid": 0
    }
}
```

**Comment by Andreagit97** [comment]
> I was checking the paths earlier and I was trying to remember how it was set up since I've not touched it in ages. I think the reason why I left it like that is because I wanted containerd to follow K3s's version as sharing them required extra configuration.
It remains bizzare how it used to work and now it doesn't.

Oh ok now it is clear, Falco 0.36.2 works also without the right container path because these `k8s.*` fields are collected by the old k8s client that is no more present in Falco 0.37.0. This also explains why both versions don't extract the `container.name` and `container.image`... the old k8s client cannot extract these container fields they are extracted only from the container runtime, but the path was wrong...

> Btw, k8smeta collector still isn't populating its fields

I don't see any `k8smeta` field in your rule. To test the plugin you need to add something to your rules, for example, some output fields like `k8s_pod_name=%k8smeta.pod.name  k8s_service_name=%k8smeta.svc.name`

**Comment by mateuszdrab** [comment]
> > I was checking the paths earlier and I was trying to remember how it was set up since I've not touched it in ages. I think the reason why I left it like that is because I wanted containerd to follow K3s's version as sharing them required extra configuration.
> > It remains bizzare how it used to work and now it doesn't.
> 
> Oh ok now it is clear, Falco 0.36.2 works also without the right container path because these `k8s.*` fields are collected by the old k8s client that is no more present in Falco 0.37.0. This also explains why both versions don't extract the `container.name` and `container.image`... the old k8s client cannot extract these container fields they are extracted only from the container runtime, but the path was wrong...
> 
> > Btw, k8smeta collector still isn't populating its fields
> 
> I don't see any `k8smeta` field in your rule. To test the plugin you need to add something to your rules, for example, some output fields like `k8s_pod_name=%k8smeta.pod.name k8s_service_name=%k8smeta.svc.name`

Shouldn't the alerts have those fields in the json though, even if the rules are not specifically using it at the moment?

**Comment by mateuszdrab** [comment]
> > Shouldn't the alerts have those fields in the json though, even if the rules are not specifically using it at the moment?
> 
> Nope at the moment you can use them just adding them to the rules

Ahh, okay.

Would be nice to see them in the log. 
I run some Loki queries and figure out what to exempt based on the most noisy and false-positive alerts. 
However, I should be fine with the basic K8s.* labels.
